training/deepspeech_training/util/sample_collections.py

```diff
@@ -570,13 +599,19 @@ def samples_from_sources(sample_sources, buffering=BUFFER_SIZE, labeled=None, re
 
     Returns
     -------
-    iterable of util.sample_collections.LabeledSample (labeled=True) or util.audio.Sample (labeled=False) supporting len
+    iterable of util.sample_collections.PackedSample if a single collection is provided, wrapping
+        LabeledSample (labeled=True) or util.audio.Sample (labeled=False) supporting len
+    or LabeledSample / util.audio.Sample directly, if multiple collections are provided
     """
     sample_sources = list(sample_sources)
     if len(sample_sources) == 0:
         raise ValueError('No files')
     if len(sample_sources) == 1:
         return samples_from_source(sample_sources[0], buffering=buffering, labeled=labeled, reverse=reverse)
-    cols = [samples_from_source(source, buffering=buffering, labeled=labeled, reverse=reverse)
+
+    # Otherwise, if we wish to interleave based on duration, we have to unpack the audio (on the fly)
+    cols = [LenMap(lambda ps: (ps.unpack() if hasattr(ps, 'unpack') else ps),
```

Comment:
This lambda is a reimplementation of `_unpack_maybe`. I'd rather move that into `sample_collections.py` and make it public, then import and use it here.

Reply:
I'd also add a comment explicitly saying that unpacking should be done lazily so that it respects the `LimitingPool` logic used in the feeding code, so that someone doesn't replace `LenMap` with a list comprehension here as a refactoring pass.

Reply:
> This lambda is a reimplementation of `_unpack_maybe`. I'd rather move that into `sample_collections.py` and make it public, then import and use it here.

Very good point. Folded in.

> I'd also add a comment explicitly saying that unpacking should be done lazily so that it respects the `LimitingPool` logic used in the feeding code, so that someone doesn't replace `LenMap` with a list comprehension here as a refactoring pass.

Added!